Source/Infrastructure/Read/ExtendedReadModelRepositoryFor.cs

```diff
@@ -0,0 +1,427 @@
+        }
+
+        public Task DeleteAsync(object id)
+        {
+            var filter = Builders<T>.Filter.Eq("_id", GetIdAsBsonValue(id));
+            return _collection.DeleteOneAsync(filter);
+        }
+
+        public Task DeleteAsync(Expression<Func<T, bool>> predicate)
+        {
+            return _collection.DeleteManyAsync(predicate);
+        }
+
+        public Task DeleteAsync(FilterDefinition<T> predicate)
+        {
+            return _collection.DeleteManyAsync(predicate);
+        }
+
+        public DeleteResult DeleteOne(object id)
+        {
+            var filter = Builders<T>.Filter.Eq("_id", GetIdAsBsonValue(id));
+            return _collection.DeleteOne(filter);
+        }
+
+        public DeleteResult DeleteOne(Expression<Func<T, bool>> predicate)
+        {
+            return _collection.DeleteOne(predicate);
+        }
+
+        public DeleteResult DeleteOne(FilterDefinition<T> predicate)
+        {
+            return _collection.DeleteOne(predicate);
+        }
+
+        public Task<DeleteResult> DeleteOneAsync(object id)
+        {
+            var filter = Builders<T>.Filter.Eq("_id", GetIdAsBsonValue(id));
+            return _collection.DeleteOneAsync(filter);
+        }
+
+        public Task<DeleteResult> DeleteOneAsync(Expression<Func<T, bool>> predicate)
+        {
+            return _collection.DeleteOneAsync(predicate);
+        }
+
+        public Task<DeleteResult> DeleteOneAsync(FilterDefinition<T> predicate)
+        {
+            return _collection.DeleteOneAsync(predicate);
+        }
+
+        public DeleteResult DeleteMany(Expression<Func<T, bool>> predicate)
+        {
+            return _collection.DeleteMany(predicate);
+        }
+
+        public DeleteResult DeleteMany(FilterDefinition<T> predicate)
+        {
+            return _collection.DeleteMany(predicate);
+        }
+
+        public Task<DeleteResult> DeleteManyAsync(Expression<Func<T, bool>> predicate)
+        {
+            return _collection.DeleteManyAsync(predicate);
+        }
+
+        public Task<DeleteResult> DeleteManyAsync(FilterDefinition<T> predicate)
+        {
+            return _collection.DeleteManyAsync(predicate);
+        }
+
+        #endregion
+
+        // This part is copied from https://github.com/dolittle-extensions/ReadModels.MongoDB/blob/master/Source/ReadModelRepositoryFor.cs
+        public BsonValue GetObjectIdFromReadModel(T readModel)
+        {
+            return GetObjectIdFrom(readModel);
+        }
+
+        private BsonValue GetObjectIdFrom(T entity)
+        {
+            try
+            {
+                var propInfo = GetIdProperty(entity);
+                object id = propInfo.GetValue(entity);
+
+                return GetIdAsBsonValue(id);
+            }
+            catch (InvalidOperationException)
+            {
+                var bsonDocument = entity.ToBsonDocument();
+                return bsonDocument["_id"];
```

Comment:
If ReadModel has property Id, that is automatically seen as _id. If not, convert it to BsonDocument and retrieve _id (assumes that we have provided a BsonClassMapping for the id for the ReadModel)

Comments, @einari ?